File: Algorithms/p109_Convert_Sorted_List_to_Binary_Search_Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        int length = 0;
        for(ListNode* p = head;p;p=p->next,++length);
        return sortedListToBSTHelper(head,length);
    }
    TreeNode* sortedListToBSTHelper(ListNode* head, int length) {
        if(length==0) return 0;
        int mid_index = length/2;
        ListNode* mid_p = head;
        for(int i=0;i<mid_index;++i){
            mid_p = mid_p->next;
        }
        TreeNode* root = new TreeNode(mid_p->val);
        root->left = sortedListToBSTHelper(head,mid_index);
        root->right = sortedListToBSTHelper(mid_p->next,length-mid_index-1);
        return root;
    }
};

File: Algorithms/p109_Convert_Sorted_List_to_Binary_Search_Tree/p109_Convert_Sorted_List_to_Binary_Search_Tree.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        length = 0
        p = head
        while p:
            p = p.next
            length += 1
        return self.sortedListToBSTHelper(head,length)

    def sortedListToBSTHelper(self, head, length):
        if length==0:
            return None
        mid_index = (length-1)//2
        mid_p = head
        for i in range(mid_index):
            mid_p = mid_p.next
        root = TreeNode(mid_p.val)
        root.left = self.sortedListToBSTHelper(head,mid_index)
        root.right = self.sortedListToBSTHelper(mid_p.next,length-mid_index-1)
        return root
